### PR TITLE
sql: avoid starvation of validation after backfill

### LINKED ISSUE #84911: bulk: large index validation after backfill can be starved by GC TTL, bounding index creation scalability
After a backfill, we run a series of validation queries (`SchemaChanger.validateIndexes`) to ensure the integrity of the new index. These amount to full table scans over the new index and the primary index.

When running an index creation on a 4TB table (32,194,851,254 rows) with a 50m GC TTL, I found that these queries could be starved and prevented from completing successfully. Instead, they would be retried every 50 minutes indefinitely with logs like the following:

```
I220722 11:59:49.467458 69188175 sql/backfill.go:2115 ⋮ [n1,job=781044268655935489,scExec,id=115,mutation=5] 250073  finished backfilling indexes
I220722 11:59:49.467505 69188175 sql/backfill.go:1398 ⋮ [n1,job=781044268655935489,scExec,id=115,mutation=5] 250074  validating new indexes
W220722 12:51:35.499342 69188175 sql/schema_changer.go:2614 ⋮ [n1,job=781044268655935489] 254915  error while running schema change, retrying: ‹verify-idx-count›: ‹batch timestamp 1658491189.517372325,0 must be after replica GC threshold 1658491263.919029492,0›
I220722 12:51:54.252033 69188175 sql/schema_changer.go:663 ⋮ [n1,job=781044268655935489,scExec,id=115,mutation=5] 254943  schema change on ‹"cash_transaction"› (v29) starting execution...
I220722 12:51:54.287300 69188175 sql/schema_changer.go:1085 ⋮ [n1,job=781044268655935489,scExec,id=115,mutation=5] 254944  stepping through state machine
I220722 12:51:54.294906 69188175 sql/schema_changer.go:1173 ⋮ [n1,job=781044268655935489,scExec,id=115,mutation=5] 254945  finished stepping through state machine
I220722 12:51:54.317960 69188175 sql/backfill.go:242 ⋮ [n1,job=781044268655935489,scExec,id=115,mutation=5] 254946  running backfill for ‹"cash_transaction"›, v=29
I220722 12:51:54.318006 69188175 sql/backfill.go:2051 ⋮ [n1,job=781044268655935489,scExec,id=115,mutation=5] 254947  backfilling 1 indexes: [‹/Table/115/{8-9}›] (writeAtRequestTimestamp: false)
I220722 12:51:54.391779 69188175 sql/backfill.go:2115 ⋮ [n1,job=781044268655935489,scExec,id=115,mutation=5] 254948  finished backfilling indexes
I220722 12:51:54.391804 69188175 sql/backfill.go:1398 ⋮ [n1,job=781044268655935489,scExec,id=115,mutation=5] 254949  validating new indexes
W220722 13:43:45.978629 69188175 sql/schema_changer.go:2614 ⋮ [n1,job=781044268655935489] 259866  error while running schema change, retrying: ‹verify-idx-count›: ‹batch timestamp 1658494314.415510637,0 must be after replica GC threshold 1658494384.053000856,0›
I220722 13:44:04.081576 69188175 sql/schema_changer.go:663 ⋮ [n1,job=781044268655935489,scExec,id=115,mutation=5] 259889  schema change on ‹"cash_transaction"› (v29) starting execution...
I220722 13:44:04.088672 69188175 sql/schema_changer.go:1085 ⋮ [n1,job=781044268655935489,scExec,id=115,mutation=5] 259890  stepping through state machine
I220722 13:44:04.090402 69188175 sql/schema_changer.go:1173 ⋮ [n1,job=781044268655935489,scExec,id=115,mutation=5] 259891  finished stepping through state machine
I220722 13:44:04.128474 69188175 sql/backfill.go:242 ⋮ [n1,job=781044268655935489,scExec,id=115,mutation=5] 259893  running backfill for ‹"cash_transaction"›, v=29
I220722 13:44:04.128532 69188175 sql/backfill.go:2051 ⋮ [n1,job=781044268655935489,scExec,id=115,mutation=5] 259894  backfilling 1 indexes: [‹/Table/115/{8-9}›] (writeAtRequestTimestamp: false)
```

At one point, I did see a single side of the validation complete just in time:
```
I220722 11:07:31.448156 788495333 sql/backfill.go:1716 ⋮ [n1,job=781044268655935489,scExec,id=115,mutation=5] 245898  validation: index ‹cash_transaction›/‹cash_transaction_ct_dts_idx› row count = 32194851254, time so far 50m55.088212272s
```
This revealed that we were scanning about 10M rows per second. Not terrible. I also confirmed that the query is using DistSQL.

There's a larger question of whether SQL execution should be creating protected timestamps whenever it sees long-running queries that might run into the GC threshold. In lieu of that, I think the schema changer should be manually placing a protected timestamp to protect these queries. Without one, we have an effective cap on the size of a table that can be given a new index.

I don't know whether this lives in Bulk or SQL Schema, so adding to both for now.

Jira issue: CRDB-17945

### PR DESCRIPTION
**Pull request by fqazi**
Fixes: #84911

Previously, after a backfill queries inside validateIndexes could be starved by GC jobs due to the lack of a protected time stamp. Hence it was possible that these queries could run into retry errors. To avoid these errors we can setup a protected timestamp, so that the GC job doesn't interfere with these queries.

Release note (bug fix): Index validation could be starved if it took longer then GC jobs for a given table. Protected timestamps are now created during index validation.

### CODE REVIEW

File: pkg/sql/backfill.go [hunk 1]
```diff
@@ -1524,71 +1540,167 @@ func (e InvalidIndexesError) Error() string {
...
+			// time.
+			var waitBeforeProtectedTS time.Duration
 			start := timeutil.Now()
+			if err := runHistoricalTxn.Exec(ctx, func(ctx context.Context, txn *kv.Txn, _ sqlutil.InternalExecutor) error {
+				_, zoneCfg, _, err := GetZoneConfigInTxn(ctx, txn, codec, tableDesc.GetID(), nil, "", true)
+				if err != nil {
+					return err
+				}
+				waitBeforeProtectedTS = ((time.Duration(zoneCfg.GC.TTLSeconds) * time.Second) *
+					indexValidationProtectTimeStampGCPct) / 100
```
Comment (stevendanna) [inline pkg/sql/backfill.go:1626]: For my own edification:

- If the zone configuration is changed after this lookup, will this wait time still be correct?

- Similarly, does running this in the historical transaction make sense?  Wouldn't we care about the most recent zone configuration?
-
- Is the cost of the protected timestamps high enough that we can't unconditionally install a PTS when starting the verification?  Is the idea that we don't want to pay that cost for small indexes that won't end up needing the PTS?

File: pkg/sql/backfill.go [hunk 2]
```diff
@@ -1524,71 +1540,167 @@ func (e InvalidIndexesError) Error() string {
...
+		protectedTSAdded.Do(func() {
+			// If we are not running a historical query, nothing to do here.
+			if runHistoricalTxn.ReadAsOf().IsEmpty() {
+				return
+			}
+			protectedtsID := uuid.MakeV4()
+			target := ptpb.MakeSchemaObjectsTarget(descpb.IDs{tableDesc.GetID()})
+			if err != nil {
+				return
+			}
```
Comment (stevendanna) [inline pkg/sql/backfill.go:1589]: `err` will always be nil at this point as nothing has written to it.

File: pkg/sql/backfill.go [hunk 3]
```diff
@@ -1728,55 +1845,146 @@ func ValidateForwardIndexes(
...
+				return protectedTSProvider.Protect(ctx, txn, rec)
+			})
+			if err != nil {
+				return
+			}
+			removeProtectedTS = func(ctx context.Context) error {
+				return db.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+					return protectedTSProvider.Release(ctx, txn, protectedtsID)
+				})
+			}
```
Comment (stevendanna) [inline pkg/sql/backfill.go:1888]: I know it isn't much code, but perhaps we could pull out a function here `protectTableForHistoricalTxn(ctx context.Context, runHistoricalTxn sqlutil.HistoricalInternalExecTxnRunner, table catalog.TableDescriptor) (func(ctx) error, error)`

File: pkg/sql/backfill_protected_timestamp_test.go [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+			ServerArgs: base.TestServerArgs{
+				Settings:                 settings,
+				DisableDefaultTestTenant: true,
+				Knobs: base.TestingKnobs{
+					SQLExecutor: &sql.ExecutorTestingKnobs{
+						BeforeExecute: func(ctx context.Context, sql string) {
+							if indexScanQuery.MatchString(sql) {
+								indexValidationQueryWait <- struct{}{}
+								<-indexValidationQueryResume
+								time.Sleep(time.Second * 5)
```
Comment (stevendanna) [inline pkg/sql/backfill_protected_timestamp_test.go:79]: What's this for? Is it to give the mvccGC queue time to process what we've enqueued.

**Review by fqazi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89540)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rhu713 and @stevendanna)

___
*[`pkg/sql/backfill.go` line 1589 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89540#-NDjLkXXc-e6IF_Kg9lu-r1-1589:-NE7gNdT8cgElHkMkvGg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aff15b7b69a7c665819c090a6f37177f48a96dd5/pkg/sql/backfill.go#L1589)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

`err` will always be nil at this point as nothing has written to it.
</blockquote></details>

Done.
___
*[`pkg/sql/backfill.go` line 1626 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89540#-NDjLkXXc-e6IF_Kg9lu-r1-1626:-NE7gOt-BWhQPIPc0Vw7:b-g0kqvk) ([raw file](https://github.com/cockroachdb/cockroach/blob/aff15b7b69a7c665819c090a6f37177f48a96dd5/pkg/sql/backfill.go#L1626)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

For my own edification:

- If the zone configuration is changed after this lookup, will this wait time still be correct?

- Similarly, does running this in the historical transaction make sense?  Wouldn't we care about the most recent zone configuration?
-
- Is the cost of the protected timestamps high enough that we can't unconditionally install a PTS when starting the verification?  Is the idea that we don't want to pay that cost for small indexes that won't end up needing the PTS?
</blockquote></details>

Good points, a couple of replies:

1) No it won't be, but it probably won't matter. We will end up retrying the job if the GC job gets to the data before we get a chance to install a protected timestamp record. If we are too fast it doesn't really matter and we just have the record early.

2) For validation the timestamp is pretty much derived from when the job started, but we can do an independent txn and get the latest. Not sure if it makes a huge difference here in terms of timing both of these things are early in the validation.
___
*[`pkg/sql/backfill.go` line 1888 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89540#-NDjLkXXc-e6IF_Kg9lu-r1-1888:-NE7gOCQ5-a4D-hxsOMh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aff15b7b69a7c665819c090a6f37177f48a96dd5/pkg/sql/backfill.go#L1888)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

I know it isn't much code, but perhaps we could pull out a function here `protectTableForHistoricalTxn(ctx context.Context, runHistoricalTxn sqlutil.HistoricalInternalExecTxnRunner, table catalog.TableDescriptor) (func(ctx) error, error)`
</blockquote></details>

Done.
___
*[`pkg/sql/backfill_protected_timestamp_test.go` line 79 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89540#-NDjLkXXc-e6IF_Kg9m2-r1-79:-NE7hOBw-8bBHp3Wsi1L:bwsa9gy) ([raw file](https://github.com/cockroachdb/cockroach/blob/aff15b7b69a7c665819c090a6f37177f48a96dd5/pkg/sql/backfill_protected_timestamp_test.go#L79)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

What's this for? Is it to give the mvccGC queue time to process what we've enqueued.
</blockquote></details>

Yes, it's a delay to allow it to process what has been enqueued.

**Review by fqazi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89540)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @rhu713, and @stevendanna)

___
*[`pkg/sql/backfill.go` line 1626 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89540#-NDjLkXXc-e6IF_Kg9lu-r1-1626:-NE7hsc8AimKCa6QXAC9:bci23i4) ([raw file](https://github.com/cockroachdb/cockroach/blob/aff15b7b69a7c665819c090a6f37177f48a96dd5/pkg/sql/backfill.go#L1626)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Good points, a couple of replies:

1) No it won't be, but it probably won't matter. We will end up retrying the job if the GC job gets to the data before we get a chance to install a protected timestamp record. If we are too fast it doesn't really matter and we just have the record early.

2) For validation the timestamp is pretty much derived from when the job started, but we can do an independent txn and get the latest. Not sure if it makes a huge difference here in terms of timing both of these things are early in the validation.
</blockquote></details>

3. Yes I think that's the case, but would like to get @ajwerner thoughts. I'm guessing stuff like multiregion makes it worse

**Review by stevendanna — COMMENTED** [review]
Thanks for taking the time to answer those questions.  I've left a few more comments on some things I noticed while reading through it again.

File: pkg/sql/backfill.go [hunk 5]
```diff
@@ -1768,55 +1906,132 @@ func ValidateForwardIndexes(
...
-			case <-tableCountsReady:
-				expectedCount := tableRowCount
-				// If the index is a partial index, the expected number of rows
-				// is different than the total number of rows in the table.
-				if idx.IsPartial() {
-					expectedCount = partialIndexExpectedCounts[idx.GetID()]
+			var indexCountReady = make(chan error)
+			var idxLenErr error
+			var idxLen int64
+			go func() {
```
[the remark this answers is not in the document]
Comment (fqazi) [reply pkg/sql/backfill.go]: Good point cleaner to just use a group context here, and avoid all the funiness.

**Review by stevendanna — APPROVED** [review]
Looks reasonable to me!

**Review by ajwerner — CHANGES_REQUESTED** [review]
I'm so sorry about the delay

> Reviewed 2 of 13 files at r1, 3 of 15 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89540#-:-NFKPXAbCTSFZVh3bkaK:b-igmcxc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @fqazi, @rhu713, and @stevendanna)

___
*[`pkg/sql/backfill.go` line 1626 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89540#-NDjLkXXc-e6IF_Kg9lu-r1-1626:-NFKP4tu8IWzeXUs_aDK:bfdy8dv) ([raw file](https://github.com/cockroachdb/cockroach/blob/aff15b7b69a7c665819c090a6f37177f48a96dd5/pkg/sql/backfill.go#L1626)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

3. Yes I think that's the case, but would like to get @ajwerner thoughts. I'm guessing stuff like multiregion makes it worse
</blockquote></details>

Above I suggest https://github.com/cockroachdb/cockroach/blob/8878715577f4b8669da2cdaa57f6a0070b784998/pkg/config/provider.go#L18, you can register for a notification if you care. I think it's better than reading it historically
___
*[`pkg/sql/backfill.go` line 1888 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89540#-NDjLkXXc-e6IF_Kg9lu-r1-1888:-NFKPK-c7M0uS2o1Dqmg:b-8rv86f) ([raw file](https://github.com/cockroachdb/cockroach/blob/aff15b7b69a7c665819c090a6f37177f48a96dd5/pkg/sql/backfill.go#L1888)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

is it done?
___
*[`pkg/sql/backfill.go` line 1648 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/89540#-NFKOavl8aZKN5yDGshb:-NFKOavl8aZKN5yDGshc:b-6pwopd) ([raw file](https://github.com/cockroachdb/cockroach/blob/d814b83e6c5352c86b5cf34c5cdade62e220cff1/pkg/sql/backfill.go#L1648)):*
> ```Go
> 		// time.
> 		var waitBeforeProtectedTS time.Duration
> 		if err := runHistoricalTxn.Exec(ctx,
> ```

My sense is that this ought to just use the in-memory system config provider. 

https://github.com/cockroachdb/cockroach/blob/8878715577f4b8669da2cdaa57f6a0070b784998/pkg/config/provider.go#L18
https://github.com/cockroachdb/cockroach/blob/70f85cd5671382a6c3843427b6deb870b3ee176f/pkg/config/system.go#L393
___
*[`pkg/sql/backfill_protected_timestamp_test.go` line 79 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89540#-NDjLkXXc-e6IF_Kg9m2-r1-79:-NFKPT4N5BxryX6mVRwW:b-ews0ch) ([raw file](https://github.com/cockroachdb/cockroach/blob/aff15b7b69a7c665819c090a6f37177f48a96dd5/pkg/sql/backfill_protected_timestamp_test.go#L79)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Yes, it's a delay to allow it to process what has been enqueued.
</blockquote></details>

Can you just run the mvccGCQueue synchronously?
___
*[`pkg/sql/sqlutil/internal_executor.go` line 241 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89540#-NE1qsZC9xQ_El77Mc7D:-NE1qsZC9xQ_El77Mc7E:bmrxen0) ([raw file](https://github.com/cockroachdb/cockroach/blob/aff15b7b69a7c665819c090a6f37177f48a96dd5/pkg/sql/sqlutil/internal_executor.go#L241)):*
> ```Go
> // HistoricalInternalExecTxnRunnerFn callback for executing with the internal executor
> // at a fixed timestamp.
> type HistoricalInternalExecTxnRunnerFn func(ctx context.Context, fn InternalExecFn) error
> ```

nit: use `=` if you aren't going to put methods on this function.
___
*[`pkg/sql/sqlutil/internal_executor.go` line 273 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89540#-NE1qjhjDvEzOEz92-2o:-NE1qjhjDvEzOEz92-2p:bn0q6y3) ([raw file](https://github.com/cockroachdb/cockroach/blob/aff15b7b69a7c665819c090a6f37177f48a96dd5/pkg/sql/sqlutil/internal_executor.go#L273)):*
> ```Go
> // executor.
> func NewHistoricalInternalExecTxnRunner(
> 	fn HistoricalInternalExecTxnRunnerFn, readAsOf hlc.Timestamp,
> ```

nit: reorder the arguments to put the timestamp first.

**Review by fqazi — COMMENTED** [review]
@ajwerner @stevendanna TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89540#-:-NFLsQMi8eqEaDHQudwt:byqns1r)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @rhu713, and @stevendanna)

___
*[`pkg/sql/backfill.go` line 1626 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89540#-NDjLkXXc-e6IF_Kg9lu-r1-1626:-NFLsPl-0DPe4qmVU1Bf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aff15b7b69a7c665819c090a6f37177f48a96dd5/pkg/sql/backfill.go#L1626)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Above I suggest https://github.com/cockroachdb/cockroach/blob/8878715577f4b8669da2cdaa57f6a0070b784998/pkg/config/provider.go#L18, you can register for a notification if you care. I think it's better than reading it historically
</blockquote></details>

Done.
___
*[`pkg/sql/backfill.go` line 1888 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89540#-NDjLkXXc-e6IF_Kg9lu-r1-1888:-NFKqoi33PFAbwfm0TuE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aff15b7b69a7c665819c090a6f37177f48a96dd5/pkg/sql/backfill.go#L1888)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

is it done?
</blockquote></details>

Done.



_Code quote:_
```Go
protectedTSInstallGrp
```
___
*[`pkg/sql/backfill.go` line 1583 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/89540#-NDjLkXXc-e6IF_Kg9lu-r2-1583:-NFKi_8c0DzXjfkwjGIa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5201a3e43ba71a5a0cc8770eb9d9d2ac7ec94c0d/pkg/sql/backfill.go#L1583)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

[nit] Since we have no naked returns in this function, I think we can removed the named return arguments here. The types make it clear what is being returned in this case.
</blockquote></details>

Done.



_Code quote:_
```Go
protectTableForHistoricalTxnFn
```
___
*[`pkg/sql/backfill.go` line 1920 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/89540#-NDjLkXXc-e6IF_Kg9lu-r2-1920:-NFKiciy93NZu_otArkP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5201a3e43ba71a5a0cc8770eb9d9d2ac7ec94c0d/pkg/sql/backfill.go#L1920)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

[nit] I think we can do the following

```suggestion
				err = errors.CombineErrors(err, removeErr)
```
</blockquote></details>

Done.
___
*[`pkg/sql/backfill.go` line 1648 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/89540#-NFKOavl8aZKN5yDGshb:-NFKiaJJ2n3jb9AGPeyZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d814b83e6c5352c86b5cf34c5cdade62e220cff1/pkg/sql/backfill.go#L1648)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

My sense is that this ought to just use the in-memory system config provider. 

https://github.com/cockroachdb/cockroach/blob/8878715577f4b8669da2cdaa57f6a0070b784998/pkg/config/provider.go#L18
https://github.com/cockroachdb/cockroach/blob/70f85cd5671382a6c3843427b6deb870b3ee176f/pkg/config/system.go#L393
</blockquote></details>

Done.
___
*[`pkg/sql/backfill_protected_timestamp_test.go` line 79 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89540#-NDjLkXXc-e6IF_Kg9m2-r1-79:-NFLsLCO8FIkbJxsH3T5:bfelp27) ([raw file](https://github.com/cockroachdb/cockroach/blob/aff15b7b69a7c665819c090a6f37177f48a96dd5/pkg/sql/backfill_protected_timestamp_test.go#L79)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Can you just run the mvccGCQueue synchronously?
</blockquote></details>

Ah, that operation was already synchronous. So, I removed the sleep completely.

**Review by ajwerner — COMMENTED** [review]
I think it's fine if you want to push back on my bookkeeping requests and save it for a later PR. Please file an issue if you do choose to defer it. One thing I would like to see is the better encapsulation of the dependencies. This sort of dependency sprawl can leave future readers wondering why this code needs access to these things. More narrow packaging of the dependencies makes future dependency injection easier. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89540#-:-NFMhr-i-X_-myy5P4Ji:bltuut4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @fqazi, @rhu713, and @stevendanna)

___
*[`pkg/sql/backfill.go` line 1577 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/89540#-NFMgSFU1FgFSjXAYE2i:-NFMgSFU1FgFSjXAYE2j:b-rlmt63) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2314070339beb3ed803437adf8fc58a29643a0d/pkg/sql/backfill.go#L1577)):*
> ```Go
> // transaction for a specific table, once a certain percentage of the GC time has
> // elapsed.
> func protectTableForHistoricalTxn(
> ```

One abstract fear I have is that if the job keeps failing because, say, nodes get killed, and then we restart, we'll be leaking these protected timestamps. Ideally we'd do something like stash the ID of the protected timestamp record in the job when we create it. I think we can do that and still backport this because, in the worst case, we lose track of the fact that we had created a protected timestamp and so we're fine. To support this, I think it'd be cool to change the signature here to be something like:

```
jobUpdateFunc func(ctx context.Context, func(context.Context, *kv.Txn, j *jobs.Job) error) error
```

And have it replace `db` and `jobID`? I think that will require the creator of the function actually having a handle to the job instance to that the leasing works. I don't know, maybe this is overkill. Another option is just to search all the protected timestamp records.

Also, let's replace the `systemConfigProvider` with something narrower like a function to get the gcttl for a descriptor ID? This way we can do better dependency injection. 

We can then use this to maybe launch a goroutine a little bit closer to the present than the ttl and see if there's a protected timestamp record already that we should just update?
___
*[`pkg/sql/backfill.go` line 1591 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/89540#-NFMeV4J0AdlVhF6SxLe:-NFMeV4J0AdlVhF6SxLf:bkqua08) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2314070339beb3ed803437adf8fc58a29643a0d/pkg/sql/backfill.go#L1591)):*
> ```Go
> 	// figure out when to apply a protected timestamp, as a percentage of this
> 	// time.
> 	zoneCfg, err := systemConfig.GetSystemConfig().GetZoneConfigForObject(codec, config.ObjectID(tableDesc.GetID()))
> ```

so, in some bad news, `GetSystemConfig` can return nil as the server is starting up. You can register to get notified of a change. I think it'd be okay to just give up if it is nil because when the job restarts in the rare case you run into the TTL, then you should find the TTL next time. 
___
*[`pkg/sql/backfill.go` line 1595 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/89540#-NFMenC450lwZ6aF7LW2:-NFMenC450lwZ6aF7LW3:b-o37dwt) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2314070339beb3ed803437adf8fc58a29643a0d/pkg/sql/backfill.go#L1595)):*
> ```Go
> 		return nil, err
> 	}
> 	waitBeforeProtectedTS := time.Duration(zoneCfg.GC.TTLSeconds) * time.Second
> ```

didn't you want to wait some factor less than 100% of the time?

**Review by fqazi — COMMENTED** [review]
> I think it's fine if you want to push back on my bookkeeping requests and save it for a later PR. Please file an issue if you do choose to defer it. One thing I would like to see is the better encapsulation of the dependencies. This sort of dependency sprawl can leave future readers wondering why this code needs access to these things. More narrow packaging of the dependencies makes future dependency injection easier.

I'm tempted to leave the last bit for a separate PR, let me get an issue open for it. But, I agree the sprawl is a mess and I think we can separate things much better by having a specific provider for setting PTS's for these cases.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89540#-:-NFOw2YoDIKInU8HL7sr:b-rfbjov)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @rhu713, and @stevendanna)

___
*[`pkg/sql/backfill.go` line 1577 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/89540#-NFMgSFU1FgFSjXAYE2i:-NFOwWxIC_xkcGd3B3Hn:b-uowdo5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2314070339beb3ed803437adf8fc58a29643a0d/pkg/sql/backfill.go#L1577)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

One abstract fear I have is that if the job keeps failing because, say, nodes get killed, and then we restart, we'll be leaking these protected timestamps. Ideally we'd do something like stash the ID of the protected timestamp record in the job when we create it. I think we can do that and still backport this because, in the worst case, we lose track of the fact that we had created a protected timestamp and so we're fine. To support this, I think it'd be cool to change the signature here to be something like:

```
jobUpdateFunc func(ctx context.Context, func(context.Context, *kv.Txn, j *jobs.Job) error) error
```

And have it replace `db` and `jobID`? I think that will require the creator of the function actually having a handle to the job instance to that the leasing works. I don't know, maybe this is overkill. Another option is just to search all the protected timestamp records.

Also, let's replace the `systemConfigProvider` with something narrower like a function to get the gcttl for a descriptor ID? This way we can do better dependency injection. 

We can then use this to maybe launch a goroutine a little bit closer to the present than the ttl and see if there's a protected timestamp record already that we should just update?
</blockquote></details>

That makes sense, I'm gonna defer it for a separate issue. I think this can be done cleanly if we fix the encapsulation issues as well.
___
*[`pkg/sql/backfill.go` line 1595 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/89540#-NFMenC450lwZ6aF7LW2:-NFOudSz5KIWK_shay_y:b6empz1) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2314070339beb3ed803437adf8fc58a29643a0d/pkg/sql/backfill.go#L1595)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

didn't you want to wait some factor less than 100% of the time?
</blockquote></details>

Good catch, this silly on my part clobbered the code accidentally.

**Comment by fqazi** [comment]
Opened https://github.com/cockroachdb/cockroach/issues/90774 for tracking

**Review by ajwerner — APPROVED** [review]
:lgtm:

> Reviewed 5 of 15 files at r2, 5 of 6 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89540#-:-NFQ5rMn72_aS-Hh2o7c:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rhu713 and @stevendanna)

**Comment by fqazi** [comment]
@ajwerner TFTR!

**Pull request merged**